text/0000-generic-atomic.md

```diff
@@ -0,0 +1,75 @@
+- Feature Name:
+- Start Date: 2016-01-22
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Add compiler support for generic atomic operations.
+
+# Motivation
+[motivation]: #motivation
+
+It is sometimes necessary or useful to use atomic operations in a generic
+context.
+
+```rust
+Atomic<T>
+```
+
+where `T` is any type. This is currently possible but fails to compile if `T` is
+too large for the native atomic operations of the target architecture.
+
+# Detailed design
+[design]: #detailed-design
+
+Four language items are added:
+
+```rust
+#[lang = "atomic_load"]
+fn atomic_load<T>(ptr: *const T, ordering: ...) -> T { ... }
+
+#[lang = "atomic_store"]
+fn atomic_store<T>(ptr: *const T, val: T, ordering: ...) { ... }
+
+#[lang = "atomic_exchange"]
+fn atomic_exchange<T>(ptr: *const T, val: T, ordering: ...) -> T { ... }
+
+#[lang = "atomic_compare_exchange"]
+fn atomic_compare_exchange<T>(ptr: *const T, old: &T, new: T, ordering: ...) -> T { ... }
```

Comment:
`compare_exchange` / `compare_and_swap`semantics are also being extended (in an orthogonal way) in #1443